core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java

```diff
@@ -29,16 +44,56 @@
 
 /**
  * @author Joe Grandja
+ * @author Madhu Bhat
  */
 public class OAuth2TokenEndpointFilter extends OncePerRequestFilter {
-	private Converter<HttpServletRequest, Authentication> authorizationGrantConverter;
+	private final String DEFAULT_TOKEN_URI = "/oauth2/token";
+	private Converter<HttpServletRequest, OAuth2AuthorizationCodeAuthenticationToken> authorizationGrantConverter;
 	private AuthenticationManager authenticationManager;
 	private OAuth2AuthorizationService authorizationService;
+	private RequestMatcher uriMatcher;
+
+	/**
+	 * Constructs an {@code OAuth2TokenEndpointFilter} using the provided parameter.
+	 *
+	 * @param authorizationService the authorization service implementation
+	 */
+	public OAuth2TokenEndpointFilter(OAuth2AuthorizationService authorizationService) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.authorizationGrantConverter = new OAuth2AuthorizationCodeAuthenticationTokenConverter();
+		this.authenticationManager = new OAuth2ClientAuthenticationManager();
+		this.uriMatcher = new AntPathRequestMatcher(DEFAULT_TOKEN_URI, HttpMethod.POST.name());
+		this.authorizationService = authorizationService;
+	}
 
 	@Override
 	protected void doFilterInternal(HttpServletRequest request,
 			HttpServletResponse response, FilterChain filterChain)
 			throws ServletException, IOException {
+		if (uriMatcher.matches(request)) {
+			String errorCode = validateAccessTokenRequest(request);
+			if (errorCode == null) {
+				OAuth2AuthorizationCodeAuthenticationToken authCodeAuthToken = authorizationGrantConverter.convert(request);
+				OAuth2AccessTokenAuthenticationToken accessTokenAuthenticationToken =
+						(OAuth2AccessTokenAuthenticationToken) authenticationManager.authenticate(authCodeAuthToken);
+				OAuth2Authorization authorization = authorizationService
+						.findByTokenAndTokenType(authCodeAuthToken.getCode(), TokenType.AUTHORIZATION_CODE);
+				authorization.setAccessToken(accessTokenAuthenticationToken.getAccessToken());
+				authorizationService.save(authorization);
```

Comment:
The access token response should be written to `response` and then `return`.

Reply:
But the issue requirements mention
> the `OAuth2AccessToken` should be updated in the in-flight `OAuth2Authorization`

So should the `OAuth2AccessToken` be updated in `OAuth2Authorization` or be written to the `response`?

Reply:
The `OAuth2AccessToken` needs to be updated in `OAuth2Authorization` AND be written to the response? I didn't specify the requirement to write to response because it's implied as per [4.1.4.  Access Token Response](https://tools.ietf.org/html/rfc6749#section-4.1.4)